greymatter-io/gitops-core
# Add RBAC Filter Toggle

To test:
create a service:
```
echo '
package greymatter

let Name= "pear"
let LocalName = "pear_local"
let EgressToRedisName = "\(Name)_egress_to_redis"

pear_config: [
  // sidecar->dashboard
  #domain & { domain_key: LocalName },
  #listener & {
    listener_key: LocalName
    _spire_self: Name
    _gm_observables_topic: Name
    _is_ingress: true
    // NOTICE ME
    _enable_rbac: true

  },
  #cluster & { cluster_key: LocalName, _upstream_port: 1337 },
  #route & { route_key: LocalName },

  // egress->redis
  #domain & { domain_key: EgressToRedisName, port: defaults.ports.redis_ingress },
  #cluster & {
    cluster_key: EgressToRedisName
    name: defaults.redis_cluster_name
    _spire_self: Name
    _spire_other: defaults.redis_cluster_name
  },
  #route & { route_key: EgressToRedisName }, // unused route must exist for the cluster to be registered with sidecar
  #listener & {
    listener_key: EgressToRedisName
    ip: "127.0.0.1" // egress listeners are local-only
    port: defaults.ports.redis_ingress
    _tcp_upstream: defaults.redis_cluster_name // NB this points at a cluster name, not key
  },

  // shared proxy object
  #proxy & {
    proxy_key: Name,
    domain_keys: [LocalName, EgressToRedisName]
    listener_keys: [LocalName, EgressToRedisName]
  },


  // edge->sidecar
  #cluster & {
    cluster_key: Name
    _spire_other: Name
  },
  #route & { domain_key: "edge", route_key: Name },

]
' >  gm/outputs/pear.cue
```

Notice the _enable_rbac field is set to true in the listener definition.

eval the cue:

```
cue eval -c -e pear_config inputs.cue gm/intermediates.cue gm/outputs/pear.cue
```
and check that the local cluster contains the the envoy_rbac filter in the active_http_filters array and in the http_filters struct:
```
{
    name:         "pear_local"
    listener_key: "pear_local"
    //...
    active_http_filters: ["gm.metrics", "gm.observables", "envoy.rbac"]
    http_filters: {
        // ...
        envoy_rbac: {
            rules: {
                action: "ALLOW"
                policies: {
                    all: {
                        permissions: [{
                            any: true
                        }]
                        principals: [{
                            any: true
                        }]
                    }
                }
            }
        }
        //...
    }
}
```

Now, to test if the mesh accepts the config. I created a fresh install of grey matter in a kops cluster called rbac-test.k8s.local.
Point your kubectl to it (make sure you are connected to tailscale):
```
awsume bs-resources-admin
kops export kubecfg rbac-test.k8s.local --admin
kubectl get svc -A
```
If kubectl get svc printed out a list of services then you connected successfully. Find and copy the edge's external domain (the very long url). 
Set the CLI's host to  `[edge external domain]:10808/services/control-api` in your config file (~/.config/greymatter/config.toml).
```
greymatter list cluster
```
to make sure you can connect.

The pear service is already deployed into the mesh. Sync the configs: 
```
greymatter sync --report --relative-path outputs --root gm cue -e pear_config
```

list the listeners:
```
greymatter list listeners | grep -A 20 -B 20 "pear_local"
```
and check if the pear_local listener contains the envoy_rbac rule.
